### DISCUSSION TITLE
pgvecto.rs 0.4.0 support

### DISCUSSION THREAD

**Discussion by diogotcorreia** [immich-app/immich#14280] [General]
Hey,

Opening this discussion just to track support for the newly released [pgvecto.rs 0.4.0](https://github.com/tensorchord/pgvecto.rs/releases/tag/v0.4.0).

We're in the process of [updating the extension in Nixpkgs](https://github.com/NixOS/nixpkgs/pull/357677).

Release notes don't appear to mention any breaking changes, but the release PR mentions a possible upgrade failure: https://github.com/tensorchord/pgvecto.rs/pull/615
Hopefully support for this is as painless as for 0.3.0.

**Comment by zackpollard** [comment]
Hey, thanks for being proactive on this. I'll leave the decision down to @mertalev but my understanding is that we aren't going to bump to 0.4.0 for pgvecto.rs due to the same team releasing the vectorchord which is better in a number of ways. They expect to suggest people migrate from pgvecto.rs to vectorchord by the end of the year, so I expect we will wait for that 🙂

https://github.com/tensorchord/VectorChord

**Comment by mertalev** [comment]
We can support 0.4.0 once they fix the upgrade issues with the release.

**Reply by dotlambda** [reply]
Weren't those fixed by https://github.com/tensorchord/pgvecto.rs/pull/617, which is included in 0.4.0?

**Reply by diogotcorreia** [reply]
@dotlambda pgvecto.rs 0.4.0 was re-released 3 times, at the time of this comment it hadn't been fixed iirc (they should've just made it a new patch release, but :man_shrugging: )

I'm mostly worried about https://github.com/tensorchord/pgvecto.rs/issues/616#issuecomment-2493285565, but I think Immich tries to recreate the index if it fails to upgrade. I'm testing it now since pgvecto.rs 0.4.0 was accidentally merged into nixpkgs, so we have to either test Immich now and patch it or revert the pgvecto.rs upgrade.

Just had a problem with my build "server" so I'm waiting for someone to physically restart it. :sweat_smile:

**Reply by mertalev** [reply]
It doesn't drop the indices before updating, just reindexes them after. [Relevant code](https://github.com/immich-app/immich/blob/6d492985a261b5d63048cc3dca936f70be822487/server/src/repositories/database.repository.ts#L100)

**Reply by diogotcorreia** [reply]
Okay, so I finally have managed to test the upgrade path with an external DB.

As expected, the direct upgrade doesn't work with the same error as in the comment linked in my previous comment. The solution is to delete the indexes:

```sql
DROP INDEX clip_index;
DROP INDEX face_index;
ALTER EXTENSION vectors UPDATE;
SELECT pgvectors_upgrade();
```

I was expecting Immich to recreate the indexes, but it's not detecting that a reindex is needed because it doesn't exist (?):

```
immich=# SELECT idx_status FROM pg_vector_index_stat WHERE indexname = 'clip_index';
 idx_status
------------
(0 rows)
```

https://github.com/immich-app/immich/blob/6d492985a261b5d63048cc3dca936f70be822487/server/src/repositories/database.repository.ts#L148

The autofix wouldn't work either way because this query also returns empty: https://github.com/immich-app/immich/blob/6d492985a261b5d63048cc3dca936f70be822487/server/src/repositories/database.repository.ts#L170

So basically to support 0.4.0 we need to find a less destructive way to get rid of the indexes or have Immich recreate them if it detects they don't exist anymore.

cc @mertalev hope that helps!

**Reply by knutknudsen** [reply]
> Okay, so I finally have managed to test the upgrade path with an external DB.
> 
> As expected, the direct upgrade doesn't work with the same error as in the comment linked in my previous comment. The solution is to delete the indexes:
> 
> ```sql
> DROP INDEX clip_index;
> DROP INDEX face_index;
> ALTER EXTENSION vectors UPDATE;
> SELECT pgvectors_upgrade();
> ```
> 
> I was expecting Immich to recreate the indexes, but it's not detecting that a reindex is needed because it doesn't exist (?):
> 
> ```
> immich=# SELECT idx_status FROM pg_vector_index_stat WHERE indexname = 'clip_index';
>  idx_status
> ------------
> (0 rows)
> ```
> 
> https://github.com/immich-app/immich/blob/6d492985a261b5d63048cc3dca936f70be822487/server/src/repositories/database.repository.ts#L148
> 
> The autofix wouldn't work either way because this query also returns empty:
> 
> https://github.com/immich-app/immich/blob/6d492985a261b5d63048cc3dca936f70be822487/server/src/repositories/database.repository.ts#L170
> 
> So basically to support 0.4.0 we need to find a less destructive way to get rid of the indexes or have Immich recreate them if it detects they don't exist anymore.
> 
> cc @mertalev hope that helps!

I am setting up immich from scratch and want to use pg17 with pgvector 0.4.0.

Is there an env to overwrite the pgvector version check?

**Reply by diogotcorreia** [reply]
@knutknudsen there doesn't seem to be anything like that, you'd have to patch it and rebuild the docker image.

This is where the check is done: https://github.com/immich-app/immich/blob/f6cbc9db06c0783d09f154f66e12d041032fff62/server/src/services/database.service.ts#L94-L96
The version range is here, if you just want to change that simply change the number: https://github.com/immich-app/immich/blob/f6cbc9db06c0783d09f154f66e12d041032fff62/server/src/constants.ts#L7

**Comment by ngosang** [comment]
VectorChord 0.1.0 was released => https://github.com/tensorchord/VectorChord/releases/tag/0.1.0
There is a blog post with impressive benchmarks => https://blog.pgvecto.rs/vectorchord-store-400k-vectors-for-1-in-postgresql

**Comment by FawenYo** [comment]
Any news on this?

**Reply by alextran1502** [reply]
We won't be supporting 0.4.0 but will be transitioning to VectorChord instead

**Reply by dani** [reply]
I bet the transition won't be painless (at least when not using the tensordchord Docker image) ...

**Reply by alextran1502** [reply]
Probably, we will also updating the db version to PG17 as well before release the stable version. 

So two birds, one stone

**Reply by the-hotmann** [reply]
@alextran1502 how this? `pgvecto.rs` in PG17 solely exists in v0.4.0 and not below: [LINK](https://hub.docker.com/r/tensorchord/pgvecto-rs/tags?name=pg17-v0.).

I also would like to see the change to VectorChord, but with a proper tutorial, since I run this in docker-compose and would like to know how to migrate.
A statusupdate and ETA would be appreciated.

**Reply by alextran1502** [reply]
We will migrate to VectorChord before hitting the stable release and upgrade to PG 17. So there will be an in-depth guide on how to migrate

**Comment by PathToLife** [comment]
VectorChord upgrade discussion here

https://github.com/immich-app/immich/discussions/17025
